File: lib/backend/src/routes/game.ts
import { Elysia, t } from 'elysia';
import { GameService } from '../services/GameService';

export const gameRoutes = new Elysia({ prefix: '/api/game' })
  .get('/logs', async () => {
    try {
      const logs = await GameService.getLatestLogContent();
      return logs;
    } catch (err: any) {
      console.error('[Game API] Error getting logs:', err);
      return '';
    }
  })
  .get('/status/running', async () => {
    try {
      const isRunning = await GameService.isGameRunning();
      const startTime = GameService.getGameStartTime();
      return { isRunning, startTime };
    } catch (err: any) {
      console.error('[Game API] Error checking running status:', err);
      return { isRunning: false };
    }
  })
  .get('/status', async () => {
    try {
      const status = await GameService.getGameStatus();
      return status;
    } catch (err: any) {
      console.error('[Game API] Error getting status:', err);
      return { 
        installed: false, 
        corrupted: false, 
        reasons: [err.message || 'Unknown error'],
        latestVersion: '0.1.0-release'
      };
    }
  })
  .get('/install/progress', () => {
    try {
      return GameService.installProgress;
    } catch (err: any) {
      return { 
        percent: 0, 
        message: err.message || 'Error retrieving progress', 
        status: 'error' 
      };
    }
  })
  .post('/install', async ({ body }) => {
    console.log('[API] /api/game/install called');
    try {
      const { version } = body as { version?: string };
      console.log(`[API] Installing version: ${version || 'latest'}`);
      const result = await GameService.installGame(version);
      console.log('[API] Install result:', result);
      return result;
    } catch (err: any) {
      console.error('[API] Install failed:', err);
      return { 
        success: false, 
        error: err.message || 'Installation failed' 
      };
    }
  }, {
    body: t.Object({
      version: t.Optional(t.String())
    })
  })
  .post('/update', async ({ body }) => {
    console.log('[API] /api/game/update called');
    try {
      const { version } = body as { version?: string };
      console.log(`[API] Updating to version: ${version || 'latest'}`);
      const result = await GameService.updateGame(version);
      return result;
    } catch (err: any) {
      console.error('[API] Update failed:', err);
      return { 
        success: false, 
        error: err.message || 'Update failed' 
      };
    }
  }, {
    body: t.Object({
      version: t.Optional(t.String())
    })
  })
  .post('/repair', async ({ body }) => {
    console.log('[API] /api/game/repair called');
    try {
      const { version } = body as { version?: string };
      console.log(`[API] Repairing game version: ${version || 'latest'}`);
      const result = await GameService.repairGame(version);
      return result;
    } catch (err: any) {
      console.error('[API] Repair failed:', err);
      return { 
        success: false, 
        error: err.message || 'Repair failed' 
      };
    }
  }, {
    body: t.Object({
      version: t.Optional(t.String())
    })
  })
  .post('/uninstall', async () => {
    console.log('[API] /api/game/uninstall called');
    try {
      const success = await GameService.uninstallGame();
      return { success };
    } catch (err: any) {
      console.error('[API] Uninstall failed:', err);
      return { 
        success: false, 
        error: err.message || 'Uninstall failed' 
      };
    }
  })
  .post('/launch', async ({ body }) => {
    try {
      const options = body as any;
      console.log(`[Game] Launch request for ${options.playerName}`);

      // Validate required fields
      if (!options.playerName || !options.uuid) {
        return {
          success: false,
          error: 'Missing required fields: playerName, uuid'
        };
      }

      if (!options.identityToken || !options.sessionToken) {
        return {
          success: false,
          error: 'Missing authentication tokens'
        };
      }

      const child = await GameService.launchGameWithFallback(options);

      return {
        success: true,
        pid: child.pid,
        message: 'Game launched successfully'
      };
    } catch (err: any) {
      console.error('[Game] Launch failed:', err);
      return {
        success: false,
        error: err.message || 'Failed to launch game'
      };
    }
  }, {
    body: t.Object({
      playerName: t.String(),
      uuid: t.String(),
      identityToken: t.String(),
      sessionToken: t.String(),
      javaPath: t.Optional(t.String()),
      gameDir: t.Optional(t.String()),
      width: t.Optional(t.Number()),
      height: t.Optional(t.Number()),
      fullscreen: t.Optional(t.Boolean()),
      server: t.Optional(t.String()),
      profileId: t.Optional(t.String()),
      gpuPreference: t.Optional(t.String())
    })
  });

File: lib/backend/src/routes/game.py
from fastapi import APIRouter, BackgroundTasks
from typing import Optional
from pydantic import BaseModel
from ..services.GameService import GameService

router = APIRouter(prefix="/api/game")

class VersionRequest(BaseModel):
    version: Optional[str] = None

class LaunchRequest(BaseModel):
    playerName: str
    uuid: str
    identityToken: str
    sessionToken: str
    javaPath: Optional[str] = None
    gameDir: Optional[str] = None
    width: Optional[int] = None
    height: Optional[int] = None
    fullscreen: Optional[bool] = None
    server: Optional[str] = None
    profileId: Optional[str] = None
    gpuPreference: Optional[str] = None

def run_install_task(version: Optional[str]):
    try:
        GameService.install_game(version)
    except Exception as e:
        print(f"Background install failed: {e}")

def run_update_task(version: Optional[str]):
    try:
        GameService.update_game(version)
    except Exception as e:
        print(f"Background update failed: {e}")

def run_repair_task(version: Optional[str]):
    try:
        GameService.repair_game(version)
    except Exception as e:
        print(f"Background repair failed: {e}")

@router.get("/status/running")
def get_running_status():
    is_running = GameService.is_game_running()
    start_time = GameService.get_game_start_time()
    return {
        "isRunning": is_running,
        "startTime": start_time
    }

@router.get("/status")
def get_status():
    return GameService.get_game_status()

@router.get("/logs")
def get_game_logs():
    return GameService.get_latest_log_content()

@router.get("/install/progress")
def get_install_progress():
    return GameService.install_progress

@router.post("/install")
def install_game(body: VersionRequest, background_tasks: BackgroundTasks):
    try:
        # Reset progress or set to initializing
        GameService.set_progress_state(0, "Initializing installation...", "installing")
        background_tasks.add_task(run_install_task, body.version)
        return {"success": True, "message": "Installation started"}
    except Exception as e:
        return {"success": False, "error": str(e)}

@router.post("/update")
def update_game(body: VersionRequest, background_tasks: BackgroundTasks):
    try:
        GameService.set_progress_state(0, "Initializing update...", "installing")
        background_tasks.add_task(run_update_task, body.version)
        return {"success": True, "message": "Update started"}
    except Exception as e:
        return {"success": False, "error": str(e)}

@router.post("/repair")
def repair_game(body: VersionRequest, background_tasks: BackgroundTasks):
    try:
        GameService.set_progress_state(0, "Initializing repair...", "installing")
        background_tasks.add_task(run_repair_task, body.version)
        return {"success": True, "message": "Repair started"}
    except Exception as e:
        return {"success": False, "error": str(e)}

@router.post("/uninstall")
def uninstall_game():
    try:
        success = GameService.uninstall_game()
        return {"success": success}
    except Exception as e:
        return {"success": False, "error": str(e)}

@router.post("/launch")
def launch_game(body: LaunchRequest):
    try:
        process = GameService.launch_game_with_fallback(body.dict())
        return {
            "success": True, 
            "pid": process.pid,
            "message": "Game launched successfully"
        }
    except Exception as e:
        return {"success": False, "error": str(e)}
